### ISSUE TITLE
ADO.NET traces missing after upgrade from 1.7 -> 1.10

### ISSUE THREAD

**Issue by bretkoppel** [DataDog/dd-trace-dotnet#581]
**Describe the bug**
After upgrading from Trace Agent 1.7 to 1.10(removing the `Datadog.Trace.ClrProfiler.Managed` package in the process), we've stopped receiving Trace data for postgres/npgsql(via Dapper). Sql Server traces are still coming through for other full framework applications on the same server. Debug logging shows the below, but never shows the "Replaced calls" message that we'd expect for the instrumented method:

`[w3wp.exe] 6288: [debug] JITCompilationStarted: function_id=140705392839240 token=100668737 name=System.Data.Common.DbCommand.System.Data.IDbCommand.ExecuteReader()`

By comparison, 1.7 shows the following:

```
[w3wp.exe] 3372: [info] *** JITCompilationStarted() replaced calls from System.Data.Common.DbCommand.System.Data.IDbCommand.ExecuteReader() to System.Data.Common.DbCommand.ExecuteDbDataReader() 100668732 with calls to Datadog.Trace.ClrProfiler.Integrations.AdoNetIntegration.ExecuteDbDataReader() 167773671
[w3wp.exe] 3372: [info] *** JITCompilationStarted() replaced calls from System.Data.Common.DbCommand.ExecuteReaderAsync() to System.Data.Common.DbCommand.ExecuteDbDataReaderAsync() 100668744 with calls to Datadog.Trace.ClrProfiler.Integrations.AdoNetIntegration.ExecuteDbDataReaderAsync() 167773672
```

**Expected behavior**
Npgsql calls via Dapper are instrumented.

**Runtime environment (please complete the following information):**
- Instrumentation mode: Have tried both automatic and with the NuGet package
- Tracer version: 1.10.0
- OS: Windows Server 2012 R2
- CLR: .NET Core 2.2

**Additional context**
- [Dapper](https://github.com/StackExchange/Dapper) 1.60.6, [Npgsql](https://github.com/npgsql/npgsql/) 4.1.1(Also tried 4.1.2)
- See also: https://github.com/DataDog/dd-trace-dotnet/issues/540

**Comment by jonjomckay** [comment]
This is also affecting us with the following environment. Previous versions did submit a few sporadic traces, but v1.10.0 now submits zero traces for database interactions.

* Instrumentation mode: Automatic
* Tracer version: 1.10.0
* OS: Linux (.NET Core Docker image)
* CLR: .NET Core 2.2

**Additional context**

* Dapper 1.60.5
* Npgsql 4.1.2

**Comment by NickAdolf** [comment]
This is an issue for us as well in tracer v1.11.0, .Net Core 2.2 on Windows.

**Comment by fehmianac** [comment]
I also same problem at dotnet core 3.1.  

I am updating my tracer v1.11.0 and removing Datadog.Trace.ClrProfiler.Managed. After that Datadog can not trace my app anymore. When i am rolling back. Everything is works.  

Instrumentation mode: Automatic
Tracer version: 1.11.0
OS: Linux (.NET Core Docker image)
CLR: .NET Core 3.1

**Comment by AndreyMarchuk** [comment]
Same issue here, downgrading to 1.9 brings the traces back

Instrumentation mode: Automatic
Tracer version: 1.10.0, 1.11.0
OS: Linux (.NET Core Docker image)
CLR: .NET Core 3.1

**Comment by bobuva** [comment]
@bretkoppel @gshackles @fehmianac @NickAdolf @AndreyMarchuk We have produced a Pre-Release [version 1.13.3-prerelease](https://github.com/DataDog/dd-trace-dotnet/releases/tag/v1.13.3-prerelease) that addresses `postgres` instrumentation through `Dapper`. Would you be willing to download it and try it out? cc: @lucaspimentel

**Comment by lucaspimentel** [comment]
We had a small issue with the versioning in this prerelease and will replace the artifacts in the github release shortly. They should be back up in a few minutes. Apologies for the inconvenience.

**Comment by lucaspimentel** [comment]
The github release is now updated with the correct installers.

**Comment by fehmianac** [comment]
While updating tracer version in Docker. New one Environment variable should be added. 
`ENV DD_DOTNET_TRACER_HOME=/opt/datadog`

After that tracer will start working. You can also remove `Datadog.Trace.ClrProfiler.Managed` package

**Comment by bretkoppel** [comment]
@bobuva @lucaspimentel I was able to give this some light testing this morning. tl;dr: 🎉Looks like we have our pg traces back!

A couple of questions:
- We're still not getting traces for our write operations, which execute through dapper's `Execute...` method([src](https://github.com/StackExchange/Dapper/blob/c1d0fd870d82d40cab2ca38be210e56245407e2f/Dapper/SqlMapper.cs#L440)). If I'm following correctly, this delegates to `DbCommand.ExecuteNonQuery...` under the hood so I'd expect it to be instrumented based on the PR. Any idea what I'm missing? This is not a regression, fwiw - The 1.7 client also didn't include write ops. I can open a separate issue if you'd prefer. 
- Is it expected that mismatches between the `DataDog.Trace` package and the installed profiler will result in e.g., `Could not load file or assembly 'Datadog.Trace, Version=1.13.3.0, Culture=neutral, PublicKeyToken=def86d061d0d2eeb'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621)`? There was a golden minute where package mismatches simply squashed tracing rather than blowing up. Just asking so that I can plan the flow for Trace Agent upgrades in services that do want code access to the Tracer.

Thanks for all of your work on this!

**Runtime Environment**
Tracer version: 1.13.3
OS: Windows Server 2012 R2
CLR: .NET Core 2.2

**Comment by bobuva** [comment]
> tl;dr: 🎉Looks like we have our pg traces back!

That's great! Thanks for the feedback!

> * We're still not getting traces for our write operations, which execute through dapper's `Execute...` method([src](https://github.com/StackExchange/Dapper/blob/c1d0fd870d82d40cab2ca38be210e56245407e2f/Dapper/SqlMapper.cs#L440)). If I'm following correctly, this delegates to `DbCommand.ExecuteNonQuery...` under the hood so I'd expect it to be instrumented based on the PR. Any idea what I'm missing? This is not a regression, fwiw - The 1.7 client also didn't include write ops. I can open a separate issue if you'd prefer.

I've done a quick scan of our code and of the calls made in `Dapper`. On the surface, I don't see why it isn't showing up. If you could create a separate issue, and if possible, include a little more context on what you're doing: async/non-async, versions, and any code you can share. cc me on that ticket and I will look into it.

> * Is it expected that mismatches between the `DataDog.Trace` package and the installed profiler will result in e.g., `Could not load file or assembly 'Datadog.Trace, Version=1.13.3.0, Culture=neutral, PublicKeyToken=def86d061d0d2eeb'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621)`? There was a golden minute where package mismatches simply squashed tracing rather than blowing up. Just asking so that I can plan the flow for Trace Agent upgrades in services that do want code access to the Tracer.

Because of the changes we've made, it is best to upgrade to the latest Tracer, which you've done, and to NOT include any of our Nuget packages. We improved the automatic instrumentation so that you do not need any Nuget packages installed manually.

Thanks Bret!

**Comment by bobuva** [comment]
> While updating tracer version in Docker. New one Environment variable should be added.
> `ENV DD_DOTNET_TRACER_HOME=/opt/datadog`
> 
> After that tracer will start working. You can also remove `Datadog.Trace.ClrProfiler.Managed` package

@fehmianac That's correct. Our documentation is also updated, indicating that you should remove that package, and add the environment variable. https://docs.datadoghq.com/tracing/setup/dotnet-core/?tab=linux

**Comment by lucaspimentel** [comment]
@bretkoppel: We released [version 1.14.0](https://github.com/DataDog/dd-trace-dotnet/releases/tag/v1.14.0) which includes the changes you tested in prerelease version above. We can now move on to #661!
